### ISSUE TITLE
Unexplained errors pushing to OSTree artifact cache

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#460]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/460)
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Jul 7, 2018, 12:33

I set up a BuildStream artifact cache and enabled pushing to it.

I got strange failures pushing, such as this one: https://gitlab.com/samthursfield/tagcloud/-/jobs/79195746

```
Failure Summary
  tagcloud-flatpak.bst:
    [00:00:00][878d7d89][ push:tagcloud-flatpak.bst          ] FAILURE Sending artifact
    [00:00:00][878d7d89][ push:tagcloud-flatpak.bst          ] FAILURE Pushing artifact
    [00:00:00][878d7d89][ push:tagcloud-flatpak.bst          ] FAILURE Failed to push artifact ['tagcloud/tagcloud-flatpak/878d7d8977edcaab0c0979fd3c276fc1354d27956b134c3250d6e089e7ef2380', 'tagcloud/tagcloud-flatpak/dc27e44748d03524f35034bdda0881b1929fb72633d67cfcad7150e8dffaf608']: Expected reply, got none

        Printing the last 20 lines from log file:
        /builds/samthursfield/tagcloud/cache/buildstream/logs/tagcloud/tagcloud-flatpak/878d7d89-push.166.log
        ======================================================================
        pushreceive: INFO: Sending update request
        pushreceive: INFO: Updating tagcloud/tagcloud-flatpak/878d7d8977edcaab0c0979fd3c276fc1354d27956b134c3250d6e089e7ef2380 0000000000000000000000000000000000000000000000000000000000000000 to a1c7b000b6743de4b6a5981ef30f3ec6e290682bc69811d613c55443af00ceec
        pushreceive: INFO: Updating tagcloud/tagcloud-flatpak/dc27e44748d03524f35034bdda0881b1929fb72633d67cfcad7150e8dffaf608 7dab26a0b89650bffda599567fe86de668689df17c2cc0bba32f6869c381cbd5 to a1c7b000b6743de4b6a5981ef30f3ec6e290682bc69811d613c55443af00ceec
        pushreceive: INFO: Enumerating objects to send
        pushreceive: INFO: Sending 12 objects
        pushreceive: INFO: Sending object 5fca4797e976fbe7761093eed33789642e0741a2b1c6cc341ba925c785191471.filez
        pushreceive: INFO: Sending object 0e7f8dd424383ca89b988b9c3c10ccef1830d498b9cee978ad67fab6f8f3637e.dirtree
        pushreceive: INFO: Sending object a1c2bc7fea480f8f67641d088677408d159cfff578517c7c35b1eda80f89df67.filez
        pushreceive: INFO: Sending object 660a2f80e630becb4e072148b0a54af085f2c24a7c093f77dbab600a4067d729.filez
        pushreceive: INFO: Sending object 6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d.dirtree
        pushreceive: INFO: Sending object 5d0d3c9558bf8873374e97d27d0e8a295c179e947562e613643be022e9f466da.dirtree
        pushreceive: INFO: Sending object 8d2c8c7bdbea21128668fce04c0a60e939021a42d9bdb47900b253a82f0a19d6.filez
        pushreceive: INFO: Sending object d9587f03e03541d05557ec06c5e90f84dd21124c6e1c151ee5e45aba7b37b94b.filez
        pushreceive: INFO: Sending object a1c7b000b6743de4b6a5981ef30f3ec6e290682bc69811d613c55443af00ceec.commit
        pushreceive: INFO: Sending object e6e38dda3d465209d0c112a58351217317b60b1b6560eeecc8b9cdf89897a724.filez
        pushreceive: INFO: Sending object 7fe2c46f1cce2d3360c196ad5ba7c98474d8d4ca216b855a3a062b1cc07f9f6d.dirmeta
        pushreceive: INFO: Sending object 1b41418fdbe5e08c87ad206d2c46e591b07f97ca7c9b67c5a4309b406da35b6a.dirtree
        [00:00:00] FAILURE [tagcloud-flatpak.bst]: Sending artifact
        [00:00:00] FAILURE [tagcloud-flatpak.bst]: Pushing artifact
        [00:00:00] FAILURE [tagcloud-flatpak.bst]: Failed to push artifact ['tagcloud/tagcloud-flatpak/878d7d8977edcaab0c0979fd3c276fc1354d27956b134c3250d6e089e7ef2380', 'tagcloud/tagcloud-flatpak/dc27e44748d03524f35034bdda0881b1929fb72633d67cfcad7150e8dffaf608']: Expected reply, got none
        ======================================================================
```

A bit of debugging showed that the problem is that the client sends a `done` message after the 'Updating' stage. I had a go at fixing this in https://gitlab.com/BuildStream/buildstream/merge_requests/532 but I don't think that's necessarily correct.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @samthursfield]](https://gitlab.com/samthursfield) on Jul 7, 2018, 12:33

mentioned in merge request !532

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jul 10, 2018, 08:30

mentioned in commit a615e00185b097a3ff3ce15f58e79a803339d9f8

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jul 10, 2018, 08:51

mentioned in commit 76cbcd4ec52c26fc77134533df64c64221610e22

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jul 10, 2018, 17:06

mentioned in commit 8c7ab2f0a3582cde19d5be7f293b47bfa51efbc4

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jul 10, 2018, 18:12

mentioned in commit 1792150b63971e2c4335f785cb9c7da1243cfcc4

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jul 16, 2018, 17:25

mentioned in commit fdbc1e33f84fd45f6ac4527846fc3cc8a14d009b

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jul 17, 2018, 06:58

mentioned in commit 1f8b4aa290a908a697f008a29ea143a9320dd639

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jul 17, 2018, 08:59

closed via merge request !337

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Jul 17, 2018, 08:59

closed via commit 1f8b4aa290a908a697f008a29ea143a9320dd639
